hal/src/common/thumbv7em/clock/v2/sources/dfll.rs

```diff
@@ -1,103 +1,344 @@
+        &self.oscctrl().dfllctrlb
+    }
+    #[allow(dead_code)]
+    #[inline]
+    fn dfllval(&self) -> &crate::pac::oscctrl::DFLLVAL {
+        &self.oscctrl().dfllval
+    }
+    #[allow(dead_code)]
+    #[inline]
+    fn dfllmul(&self) -> &crate::pac::oscctrl::DFLLMUL {
+        &self.oscctrl().dfllmul
+    }
+    #[allow(dead_code)]
+    #[inline]
+    fn dfllsync(&self) -> &crate::pac::oscctrl::DFLLSYNC {
+        &self.oscctrl().dfllsync
+    }
+    #[allow(dead_code)]
+    #[inline]
+    fn wait_sync_enable(&self) {
+        while self.dfllsync().read().enable().bit() == true {}
+    }
+    #[allow(dead_code)]
+    #[inline]
+    fn wait_sync_dfllmul(&self) {
+        while self.dfllsync().read().dfllmul().bit() == true {}
+    }
+    #[allow(dead_code)]
+    #[inline]
+    fn wait_sync_dfllval(&self) {
+        while self.dfllsync().read().dfllval().bit() == true {}
+    }
+    #[allow(dead_code)]
+    #[inline]
+    fn wait_sync_dfllctrlb(&self) {
+        while self.dfllsync().read().dfllctrlb().bit() == true {}
+    }
+    #[allow(dead_code)]
+    #[inline]
+    fn enable(&mut self) {
+        self.dfllctrla().modify(|_, w| w.enable().set_bit());
+        self.wait_sync_enable();
+    }
+    #[allow(dead_code)]
+    #[inline]
+    fn set_open_mode(&mut self) {
+        self.dfllctrlb().modify(|_, w| w.mode().clear_bit());
+        self.wait_sync_enable();
+    }
+    #[allow(dead_code)]
+    #[inline]
+    fn set_closed_mode(&mut self) {
+        self.dfllctrlb().modify(|_, w| w.mode().set_bit());
+        self.wait_sync_enable();
+    }
+    #[allow(dead_code)]
+    #[inline]
+    fn set_fine_maximum_step(&mut self, value: u8) {
+        self.dfllmul()
+            .modify(|_, w| unsafe { w.fstep().bits(value) });
+        self.wait_sync_dfllmul();
+    }
+    #[allow(dead_code)]
+    #[inline]
+    fn set_coarse_maximum_step(&mut self, value: u8) {
+        self.dfllmul()
+            .modify(|_, w| unsafe { w.cstep().bits(value) });
+        self.wait_sync_dfllmul();
+    }
+    #[allow(dead_code)]
+    #[inline]
+    fn set_multiplication_factor(&mut self, value: u16) {
+        self.dfllmul().modify(|_, w| unsafe { w.mul().bits(value) });
+        self.wait_sync_dfllmul();
+    }
+}
+
+type MultiplicationFactor = u16;
+type CoarseMaximumStep = u8;
+type FineMaximumStep = u8;
+type Fine = u8;
+type Coarse = u8;
+
+pub trait Mode: Sealed {}
+
+pub struct OpenMode {
+    // TODO: Add support for custom fine and coarse? Otherwise remove it.
+    #[allow(dead_code)]
+    fine: Option<Fine>,
+    #[allow(dead_code)]
+    coarse: Option<Coarse>,
+}
+impl Mode for OpenMode {}
+impl Sealed for OpenMode {}
+pub struct ClosedMode<T: PclkSourceType> {
+    reference_clk: Pclk<Dfll48, T>,
+    coarse_maximum_step: CoarseMaximumStep,
+    fine_maximum_step: FineMaximumStep,
+}
+impl<T: PclkSourceType> Sealed for ClosedMode<T> {}
+impl<T: PclkSourceType> Mode for ClosedMode<T> {}
+
+pub struct DfllConfig<TMode: Mode> {
+    regs: Registers,
```

Comment:
We should settle on consistent naming here. I think you don't like `token`?

I wouldn't be opposed to `regs` if we were making the `Registers` API public. But I don't think that's a good idea, as I've said before. To do so, I think we would have to consider a lot more possible interactions between the PAC-like API in `Registers` and the HAL API. It's also more public API surface to maintain, but I don't think people will actually use it. If someone really wants PAC-like access, I say let them use the top-level escape hatch provided with the `PacClocks` struct at the root of `clock::v2`.

If not `regs`, then I suggest `token` for a few reasons. Token implies two things: (1) that the type has no utility on its own and (2) that it can be exchanged for something that does have utility. It's just like tokens at an arcade. Do you have any other suggestions?

Reply:
No no, it was purely my negligence with regard to naming. I fully agree with your idea of tokens. Fixed.

Reply:
One more topic to discuss, related to this.

Right now I have `Token` as a type alias for the `Registers` struct. But I don't particularly like that. That was actually my fallback option after *really* not liking my first approach.

At first, I created independent `Token` structs that took ownership of a `Registers` struct. Then the `Config` struct (or equivalent) took ownership of the `Token`. And finally, the outermost type took ownership of the `Config`. So at the top level, to access the registers, I had code like

```rust
self.config.token.regs.method()
```

It was really verbose and I hated it. I used the `Token` alias to eliminate one layer.

```rust
self.config.token.method()
```

However, it doesn't make much sense to name something `token` but use it as registers.

Maybe the solution is to create a wrapper `Token` struct that owns a `Registers` struct. But the `Config` struct unpacks the `Token` and takes out the `Registers`. Then you would get

```rust
self.config.regs.method()
```
Then, when freeing the clock, you would recreate a `Token`.

That feels better to me. Thoughts?

Reply:
Wrapper `Token` for `Registers` that gets unpacked upon construction makes sense to me.